src/OpenFeature.DependencyInjection/OpenFeatureBuilder.cs

```diff
@@ -0,0 +1,17 @@
+using Microsoft.Extensions.DependencyInjection;
+
+namespace OpenFeature;
+
+/// <summary>
+/// Describes a <see cref="OpenFeatureBuilder"/> backed by an <see cref="IServiceCollection"/>.
+/// </summary>
+/// <param name="Services">The <see cref="IServiceCollection"/> instance.</param>
+public sealed record OpenFeatureBuilder(IServiceCollection Services)
```

Comment:
@lukas-reining Thanks for your review.

**Reasoning for using a `record` in `OpenFeatureBuilder` and extension methods**:

1. **Immutability**:  
   Using a `record` in `OpenFeatureBuilder` ensures immutability, which is crucial in configurations where state changes should not happen after initialization. Records in C# provide an easy way to define immutable types and handle value-based equality. This matches the pattern used in other parts of ASP.NET Core, like the [AuthenticationBuilder](https://github.com/dotnet/aspnetcore/blob/main/src/Security/Authentication/Core/src/AuthenticationBuilder.cs) class, where the object holds configuration data (like `IServiceCollection`) and doesn’t change its internal state directly.

2. **Following .NET Patterns (e.g., AuthenticationBuilder)**:  
   The `AuthenticationBuilder` is a great example from ASP.NET Core, where the builder encapsulates `IServiceCollection` and allows adding services via extension methods. Our approach to `OpenFeatureBuilder` follows the same design, where extension methods add specific configuration options, making the builder pattern more extensible without altering its core structure. Microsoft actively uses this builder pattern across many of their packages, not just for authentication. It's proven to be a robust way to add functionality in a clean and modular manner.

3. **Extension Methods Simplify Extensibility**:  
   Like how [GoogleExtensions.cs](https://github.com/dotnet/aspnetcore/blob/main/src/Security/Authentication/Google/src/GoogleExtensions.cs) and other authentication schemes are added via extension methods on `AuthenticationBuilder`, the same idea applies to `OpenFeatureBuilder`. This allows developers to extend the functionality of `OpenFeatureBuilder` in a modular way, adding new features without needing to modify or subclass the builder itself. This aligns with Microsoft’s general approach to keep core objects clean and extensible via extensions.

4. **Many Uses Across Microsoft Packages**:  
   While `AuthenticationBuilder` is a useful example, this builder and extension pattern is widely used throughout Microsoft’s packages, such as in DI (Dependency Injection), logging, and configuration. It allows developers to write clean, easily maintainable code that is open to extensions without modifying the core framework.

5. **Reconsidering the `sealed` Keyword**:  
   In some cases, the use of `sealed` on a builder might restrict extensibility via inheritance. Although the builder pattern usually focuses on extension methods rather than inheritance, removing `sealed` can be beneficial if you envision a need for developers to extend `OpenFeatureBuilder` through inheritance. It opens up the possibility for advanced users to create custom builders while still maintaining the base functionality.

---

In summary, the use of a `record` for `OpenFeatureBuilder` fits well with established patterns in .NET, like `AuthenticationBuilder`, making it immutable and easily extensible through extension methods. The use of this pattern is not isolated; it’s widespread across Microsoft packages and proves to be a useful approach. The consideration of removing `sealed` can further enhance the extensibility of the builder, allowing more advanced use cases where inheritance might be needed.

Reply:
I will remove the `sealed` modifier.

Reply:
Thanks @arttonoyan!
I am still not sure why we use the record. 

> Using a record in OpenFeatureBuilder ensures immutability, which is crucial in configurations where state changes should not happen after initialization.

If I am not looking at the wrong place, we do not have any data. 
The `AuthenticationBuilder` is a `class` and not a `record`. 

The `AuthenticationBuilder` also provides the basic functionality directly so maybe we would only need the extensions from [src/OpenFeature.Hosting/OpenFeatureBuilderExtensions.cs](https://github.com/open-feature/dotnet-sdk/pull/310/files#diff-f7f57b52b713d8e46a96e67b93db71b27c6c2af857daf29763105ad1a3d68e83) but we could inline [src/OpenFeature.DependencyInjection/OpenFeatureBuilderExtensions.cs](https://github.com/open-feature/dotnet-sdk/pull/310/files#diff-d0b3eb2095f69d2fad489ea905de2729b3022c21149eaf6f9c3de9236d0e17f0) as they are directly defined in the same package. 

I have no hard opinion on this, I would just like to understand the reasoning.

Reply:
I think in our case, there isn't a significant difference between using a class or a record. The main reason I opted for a record is because it allows for a more concise and compact syntax, especially when the class primarily holds immutable data, like the Services property. Functionally, it would be nearly identical to using a class like this:
```
public class OpenFeatureBuilder
{
    public OpenFeatureBuilder(IServiceCollection services)
    {
        Services = services;
    }

    public IServiceCollection Services { get; }
}
```

Reply:
> especially when the class primarily holds immutable data, like the Services property.

In our case we have one immutable property `Services` and one mutable `IsContextConfigured`.
This is not too important I guess but I am just wondering why we should do different then e.g. `ASP.NET Core` in your example.

As you said you removed `sealed` for extensibility, but I would argue that classes, as for `AuthenticationBuilder `, would be more idiomatic than extending a record as "a class can not inherit from a record" and the pattern for extending OpenFeature would be different than for the `ASP.NET Core` built-ins: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/builtin-types/record#inheritance

Reply:
For me the use of `record` also caused confusion. I had to read a bit to understand why we'd use it in this case.

Personally, since other builders we have are implemented as classes I think I'd prefer that, but it's not a strong preference.

Reply:
@lukas-reining @toddbaert Thank you for your thoughtful feedback! I completely understand your point, especially regarding consistency with other builders being implemented as classes. I can see how using a record in this case might cause some confusion. It’s not a strong preference for me either, I will change with the more familiar pattern of using classes for builders.

Reply:
Updated

Reply:
To add, other reason why we should seal classes is in dotnet 6, we got some small performance improvements: https://devblogs.microsoft.com/dotnet/performance-improvements-in-net-6/#“peanut-butter”

As for the record, I would prefer to start using it by the reasons that @arttonoyan mentioned before. But I am happy to keep as a class until we decide to move it to record.